Диплом 3 UI тесты

EvgenyYAgit opened the pull request:
Написаны тесты, запущены, пройдены, сгенерирован отчет.

EvgenyYAgit committed:
Коммит №1
Написаны тесты, запущены, пройдены, сгенерирован отчет.

AlexKitIt commented on data/data.py:
```diff
@@ -0,0 +1,65 @@
+    headers = {
+        'Authorization': f'{token}'
+    }
+    requests.delete("https://stellarburgers.nomoreparties.site/api/auth/user", headers=headers)
```
Нужно исправить: здесь и далее: урлы стоит хранить отдельно, они постоянно переиспользуются

AlexKitIt commented on data/data.py:
Нужно исправить: в data хранятся только статичные данные, судя по наполнения модуль должен называться  helpers

AlexKitIt commented on data/data.py:
```diff
@@ -0,0 +1,65 @@
+
+
+# удаление пользователя по api
+def api_user_delete(token):
```
Нужно исправить: здесь и далее: над каждым логическим методом ставится аннотация allure.step

AlexKitIt commented on data/data.py:
```diff
@@ -0,0 +1,65 @@
+    token = headers
+    user_email = data_user['email']
+    user_password = data_user['password']
+    driver.click_on_section(locators.personal_area_locators.personal_area)
+    driver.wait_element_located(locators.personal_area_locators.entrance_text)
+    driver.input_text(locators.personal_area_locators.string_input_email, user_email)
+    driver.input_text(locators.personal_area_locators.string_input_password, user_password)
+    driver.click_on_section(locators.personal_area_locators.entrance_button)
+    return token
+
+
+# переход к восстановлению пароля
+def restore_password(driver):
+    driver.click_on_section(locators.password_recovery_locators.login_button)
+    driver.wait_element_clickable(locators.password_recovery_locators.restore_password)
+    driver.click_on_section(locators.password_recovery_locators.restore_password)
```
Нужно исправить: ui методов в данном проекте зашитых в хелперы быть не может. Только апи. Все методы предусловий на ui описываются в conftest с использованием pom

AlexKitIt commented on locators/basic_functionality_locators.py:
```diff
@@ -0,0 +1,29 @@
+
+
+# кнопка закрыть окно ингредиента
+exit_button = [By.XPATH, "/html/body/div[1]/div/section[1]/div[1]/button"]
```
Нужно исправить: здесь и далее: использование порядкового номера элемента в dom является весьма нестабильным способом поиска, необходимо привязываться к атрибутам конкретного элемента
Кроме того ходить от html и использовать абсолютные пути нельзя - очень ненадежно, добавится на пути новый элемент - тест ляжет

AlexKitIt commented on pages/home_page.py:
```diff
@@ -0,0 +1,61 @@
+from seletools.actions import drag_and_drop
+
+
+class HomePage:
```
А где здесь pom??
Объект страницы - это одна конкретная страница. А в этих объектах мы храним конкретные элементы и описываем действие с ними. 
Кроме того в работе просто остутствует BasePage, в котором будут базовые методы для взаимодействия со всеми элементами по типу клика или send_keys и базовые wait, которых уже не должно быть в объектах страниц, 
там, за счет наследования объектов от Base, можно будет эти общие методы вызывать уже для элементов страниц. Вот эта модель и будет PageObject. 
Все обращения к driver - в BasePage, все работа с конкретными элементами на страницах - в объектах страниц

AlexKitIt commented on pages/home_page.py:
```diff
@@ -0,0 +1,61 @@
+
+    # переход на страницу
+    def get_site(self, site):
+        self.driver.get(site)
```
Нужно исправить: здесь и далее по проекту: все прямые обращения к driver и waitы необходимо вынести в base и оттуда за счет наследования переиспользовать в объектах страниц

AlexKitIt commented on pages/home_page.py:
```diff
@@ -0,0 +1,61 @@
+        self.driver = driver
+
+    # переход на страницу
+    def get_site(self, site):
```
Нужно исправить: здесь и далее: над каждым методом пейджи ставится аннотация allure.step

AlexKitIt commented on tests/test_basic_functionality.py:
```diff
@@ -0,0 +1,76 @@
+    def test_click_constructor(self, driver):
+        some_object = HomePage(driver)
+        some_object.get_site(data.variables.site)
+        some_object.wait_element_clickable(locators.basic_functionality_locators.order_feed_button)
```
Нужно исправить: здесь и далее: локаторов в тестах быть не должно, если это не вариант параметризации, они живут в методах пейджи, тест должен быть максимально изолирован от логики выполнения

AlexKitIt commented on tests/test_basic_functionality.py:
```diff
@@ -0,0 +1,76 @@
+
+    @allure.title('Переход по клику на «Конструктор»')
+    def test_click_constructor(self, driver):
+        some_object = HomePage(driver)
```
Нужно исправить: страниц в тестах практически всегда будет создаваться несколько, это е2е сценарии

EvgenyYAgit committed:
Коммит №2
urls теперь хранятся отдельно в папке data файл urls

EvgenyYAgit committed:
Коммит №3
файл data переименован в более подходящее название helpers так-как хранит только статичные данные

EvgenyYAgit committed:
Коммит №4
Над каждым логическим методом установлена аннотация allure.step

EvgenyYAgit committed:
Коммит №5
Улучшена стабильность моззилы засчет настройки full hd разрешения браузера.
UI методы убраны из helpers.

EvgenyYAgit committed:
Коммит №6
Исправлен локатор который исользовался с порядковыми номерами элемента и начинался с html.

EvgenyYAgit committed:
Коммит №7
Проект переписан под модель PageObject.
Создан родительский метод BasePage.
Все обращения к driver в BasePage.
Над каждым методом проставлена аннотация allure.step.
Все локаторы убраны в пейджи по своему функционалу и тестам.
Кросплатформенно создаются несколько страниц для е2е сценариев.
Все тесты успешно проходят.

AlexKitIt commented on data/helpers.py:
```diff
@@ -0,0 +1,59 @@
+
+@allure.step('Создание пользователя по api, логин и возврат токена для удаления')
+def login_unique_user(driver):
+    user_data = generate_unique_data()
```
Можно лучше: если мы используем этот метод только один раз, почему мы сразу не возвращаем из него  payload = {
        "email": f'{user_data[0]}',
        "password": f'{user_data[1]}',
        "name": f'{user_data[2]}'
    }?

AlexKitIt commented on data/helpers.py:
```diff
@@ -0,0 +1,59 @@
+    token = headers
+    user_email = data_user['email']
+    user_password = data_user['password']
+    driver.click_on_section(locators.personal_area_locators.personal_area)
+    driver.wait_element_located(locators.personal_area_locators.entrance_text)
+    driver.input_text(locators.personal_area_locators.string_input_email, user_email)
+    driver.input_text(locators.personal_area_locators.string_input_password, user_password)
+    driver.click_on_section(locators.personal_area_locators.entrance_button)
```
Нужно исправить: в этом проекте везде, где есть действия на ui использовать pom обязательно, а значит не может быть прямых обращений к driver за пределами basePage
Кроме того, не надо объединять методы апи и ui в одной функции. Оставь тут апи, а для ui напиши отдельную фикстуру

AlexKitIt commented on pages/basic_page.py:
```diff
@@ -0,0 +1,35 @@
+
+    @allure.step('Ожидание элемента и нажатие через скрипт')
+    def wait_and_click(self, element):
+        wait_element = WebDriverWait(self.driver, 30).until(expected_conditions.element_to_be_clickable(element))
```
Нужно исправить: метод описан в wait_element_clickable необходимо переиспользовать а не копипастить

AlexKitIt commented on pages/basic_functionality_page.py:
```diff
@@ -0,0 +1,93 @@
+
+    @allure.step('Перенос элемента')
+    def drag_and_drop_element(self):
+        source = self.driver.find_element(*locators.basic_functionality_locators.ingredient_button)
+        target = self.driver.find_element(*locators.basic_functionality_locators.place_for_ingredient)
+        drag_and_drop(self.driver, source, target)
+
+    @allure.step('Ввод сообщения в строку input')
+    def input_text(self, string, message):
+        self.driver.find_element(*string).send_keys(message)
```
Нужно исправить: повторный коммент: здесь и далее по проекту: все прямые обращения к driver  необходимо вынести в base и оттуда за счет наследования переиспользовать в объектах страниц

AlexKitIt commented on pages/basic_functionality_page.py:
```diff
@@ -0,0 +1,93 @@
+
+    @allure.step('Переход в Ленту заказов')
+    def click_on_order_feed(self):
+        self.wait_element_clickable(locators.basic_functionality_locators.order_feed_button)
```
Можно лучше: locators.basic_functionality_locators стоит записать отдельно в атрибут класса, тогда в методах будут просто аккуратные self.order_feed_button

AlexKitIt commented on pages/section_order_feed_page.py:
```diff
@@ -0,0 +1,126 @@
+        WebDriverWait(self.driver, 30).until(lambda driver: driver.find_element(
+            *locators.section_order_feed_locators.number_order).text != data.variables.default_number)
+        # возврат нового текста
+        new_text = self.driver.find_element(*locators.section_order_feed_locators.number_order).text
```
Нужно исправить: повторный коммент: здесь и далее по проекту: все прямые обращения к driver и waitы необходимо вынести в base и оттуда за счет наследования переиспользовать в объектах страниц

AlexKitIt commented on pages/section_order_feed_page.py:
```diff
@@ -0,0 +1,126 @@
+            *locators.section_order_feed_locators.number_order).text != data.variables.default_number)
+        # возврат нового текста
+        new_text = self.driver.find_element(*locators.section_order_feed_locators.number_order).text
+        print(new_text)
```
Нужно исправить: принтов быть не должно

AlexKitIt commented on locators/personal_area_locators.py:
```diff
@@ -0,0 +1,24 @@
+# кнопка войти
+entrance_button = [By.XPATH, "//button[text()='Войти']"]
+# текст соберите бургер
+assemble_the_burger_text = [By.XPATH, "//h1[text()='Соберите бургер']"]
```
Нужно исправить: везде по локаторам: копипасты локаторов быть не должно, элеменет всегда находится только на одной странице, поэтому его локатор не надо описывать в других пейджах. Так нарушается pom. Е2Е тест вообще редко выглядит как создание объекта одной страницы и работы с ней, поэтому всегда бросается в глаза. В тесте должен создаваться объект страницы - выполняться действия на ней, потом создается объект другой страницы и выполняются действия на ней

EvgenyYAgit committed:
Коммит №8
Метод генерации данных пользователя generate_unique_data возвращает сразу payload

EvgenyYAgit committed:
Коммит №9
Все driver удалены, вместо них добавлены в page недостающие действия. Все действия в ui организованы черезе

EvgenyYAgit committed:
Коммит №10
Все действия в ui организованы через pom. Убран неиспользуемый метод.

EvgenyYAgit committed:
Коммит №11
локаторы записаны в отдельный атрибут классов и вызываются через self

EvgenyYAgit committed:
Коммит №12
Локаторы приведены к POM. Добавлен родительский локатор, остальные стали дочерними. В родительском локаторе общие локаторы с главной страницы сайта.
Тесты проходят успешно.

AlexKitIt commented on data/helpers.py:
```diff
@@ -0,0 +1,64 @@
+
+@allure.step('Логин сгенерированного пользователя')
+def login(page, user_email, user_password):
+    page.click_on_section(locators.personal_area_locators.PersonalAreaLocators.personal_area)
```
У нас же известно какая именно пейджа тут будет, зачем метод ее принимает?
